From 2d5ccf42dc082a8fcaa5df9fa565b8ef2c0c9ea7 Mon Sep 17 00:00:00 2001
From: Boris Fiuczynski <fiuczy@linux.ibm.com>
Date: Mon, 19 Nov 2018 10:09:51 +0100
Subject: [PATCH 5/6] qemu: domain device definition hostdev validation

This is code taken from commit d54e45b6edd7 in preparation for
the hostdev validation refactoring.

Signed-off-by: Boris Fiuczynski <fiuczy@linux.ibm.com>

Backport-Note: reduced a bigger rework to just what is needed to expose
this feature.
Author: Boris Fiuczynski <fiuczy@linux.ibm.com>
Original-Author: Erik Skultety <eskultet@redhat.com>
Origin: backport, https://libvirt.org/git/?p=libvirt.git;a=commit;h=d54e45b6edd7623e488a19e30bc4148a21fa8b03
Bug-Ubuntu: https://bugs.launchpad.net/bugs/1787405
Last-Update: 2018-11-20
---
 src/qemu/qemu_domain.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/src/qemu/qemu_domain.c b/src/qemu/qemu_domain.c
index 4de5801..555dcd4 100644
--- a/src/qemu/qemu_domain.c
+++ b/src/qemu/qemu_domain.c
@@ -3891,6 +3891,24 @@ qemuDomainDeviceDefValidateHostdev(const virDomainHostdevDef *hostdev,
         return -1;
     }
 
+    if (hostdev->mode == VIR_DOMAIN_HOSTDEV_MODE_SUBSYS) {
+        switch ((virDomainHostdevSubsysType) hostdev->source.subsys.type) {
+        case VIR_DOMAIN_HOSTDEV_SUBSYS_TYPE_USB:
+        case VIR_DOMAIN_HOSTDEV_SUBSYS_TYPE_PCI:
+        case VIR_DOMAIN_HOSTDEV_SUBSYS_TYPE_SCSI:
+        case VIR_DOMAIN_HOSTDEV_SUBSYS_TYPE_SCSI_HOST:
+        case VIR_DOMAIN_HOSTDEV_SUBSYS_TYPE_MDEV:
+            break;
+        case VIR_DOMAIN_HOSTDEV_SUBSYS_TYPE_LAST:
+        default:
+            virReportError(VIR_ERR_INTERNAL_ERROR,
+                           _("Unexpected enum value %d for "
+                             "virMediatedDeviceModelType"),
+                           hostdev->source.subsys.type);
+            return -1;
+        }
+    }
+
     return 0;
 }
 
-- 
2.17.0

